### PR TITLE
Rename "layout" to "kynd"

### PR DESCRIPTION
**Pull request by goldfirere**
In the brave new modal kinds world, types are classified by *kinds*, not *layouts*. A layout is, instead, just one component of a kind.

The word kind is overloaded in the source. It's already in use in the type-checker, as part of a type declaration. To avoid confusion (and for easy recognition when editing code), this patch does not use `kind` in the source, but instead uses `kynd`. Perhaps one day we will convince upstream to rename type declaration kinds to something else, and then we can replace `kynd` with `kind`. But until that day, `kynd` it is.

(There was a discussion some time ago about this, and `kynd` won out over `kkind`, which was my original attempt.)

Two other changes in this PR:
* I restructured (slightly) the module structure in kynd.ml (originally layouts.ml).
* I clarified an error message by a tiny bit as I swooped by.

### CODE REVIEW

**Comment by goldfirere** [comment]
Changed to use jkind, by popular request

File: ocaml/lambda/matching.ml
Comment (ccasin) [inline ocaml/lambda/matching.ml:102]: I'm surprised to see this error change.  I think the error is really about layouts.  We might add other similar errors in the future for the other parts of `Jkind`, if there are particularly experimental ones, but I don't think this would grow to encompass them.

File: ocaml/lambda/matching.ml
Comment (ccasin) [inline ocaml/lambda/matching.ml:111]: Similarly I think this check is really about layouts.

File: ocaml/lambda/translcore.ml
Comment (ccasin) [inline ocaml/lambda/translcore.ml:32]: Another error I think is really about layouts.  (That is, I think we wouldn't have trouble compiling probes that capture things where the other bits of the jkind are unusual).

File: ocaml/typing/typeopt.ml
Comment (ccasin) [inline ocaml/typing/typeopt.ml:25]: As in the `transl` files, my intuition is that this error is properly about layouts.

File: ocaml/typing/typeopt.ml
Comment (ccasin) [inline ocaml/typing/typeopt.ml:199]: Probably this function really takes a layout argument and shouldn't be renamed.

File: ocaml/typing/typeopt.ml
Comment (ccasin) [inline ocaml/typing/typeopt.ml:394]: Stray formatting changes?

File: ocaml/typing/typeopt.ml
Comment (ccasin) [inline ocaml/typing/typeopt.ml:427]: I think these comments are really about layouts.  (Though probably there are a lot of such comments, and it's not worth fixing them all)

**Review by ccasin — APPROVED** [review]
To record some out-of-band discussion:  Richard and I disagree a bit about the ideal structure of this PR.  Particularly in the new `Jkind.ml`, I think it would be clearer to leave the `Jkind.Layout` module unchanged, and jkinds should be added as a new thing `Jkind.t` that will today be an alias of `Jkind.Layout.t` but grow some new fields over time.  Richard prefers it the way he's done it here.

I don't think it's worth holding this up to argue further, though, so I'm approving.  To mitigate one major downside of Richard's approach, he's going to break this into two PRs.  The first just renames the file, and the second does the rest of these changes.  That way the history will record the file name change as a move.

I followed the suggestion review procedure.  I focused my spot check on files around the edges of the typechecker, and have some minor suggestions in those files about things I think should keep their old name (but feel free to disagree).

File: ocaml/testsuite/tests/typing-layouts/datatypes.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-layouts/datatypes.ml:6]: Almost all of the additions to type-checking are about kinds, I think, not layouts. The only real exception is around the checks to make sure that a type is representable (that is, not `any`). But any check for, say, compatbility between a declared classifier on a type and some other type is really a kind check, not just a layout check.

I went through the testsuite files and moved some of the comments back to layout.

File: ocaml/lambda/matching.ml
Comment (goldfirere) [reply ocaml/lambda/matching.ml:102]: Yes, I agree here. I've reverted this change.

File: ocaml/lambda/matching.ml
Comment (goldfirere) [reply ocaml/lambda/matching.ml:111]: This is debatable. The thing we care about is whether the layout is value, as you say. But the argument to this function must remain a jkind, because that's what we have to pass it. That all might change, but building up that distinction is going to happen in future PRs; I do not want to implement that in this one. I'll see if I can find a better name.

Comment (ccasin) [reply ocaml/lambda/matching.ml:111]: (I'm convinced, no need to find a better name.)

File: ocaml/lambda/translcore.ml
Comment (goldfirere) [reply ocaml/lambda/translcore.ml:32]: OK

File: ocaml/typing/typeopt.ml
Comment (goldfirere) [reply ocaml/typing/typeopt.ml:199]: Disagree here. Types are classified by jkinds, not layouts, and the arguments to this function are type classifiers.

File: ocaml/typing/typeopt.ml
Comment (goldfirere) [reply ocaml/typing/typeopt.ml:394]: Yup. Reverted.

File: ocaml/typing/typeopt.ml
Comment (goldfirere) [reply ocaml/typing/typeopt.ml:427]: Yeah -- I actually tend to agree with you here. I've looked for uses of jkind in this file and changed as I saw appropriate.

**Comment by goldfirere** [comment]
This PR has now been restructured to depend on #1886. This means git can recognize that what was previously layouts.ml is now jkind.ml.

I have resolved a few Discussions above, as Chris elsewhere invited me to do. I think once CI passes on #1886, we can merge that, and then this.

**Comment by goldfirere** [comment]
The current state here is a little inconsistent: With this patch, the type checker loses its notion of "layout". Instead, it just has jkinds. But the modal kinds work will introduce a new concept called a layout. And so @ccasin wanted some comments to jump to their end state of saying "layout" instead of "jkind", even though I claim that the latter would be more consistent. Using "layout" means fewer changes, and the inconsistency will be temporary. An alternative approach would be to add the (re)introduction of layouts into this PR, but I thought having this be just a renaming -- no new logic! -- would be better. I think there's no perfect answer here.

**Pull request merged**
